File: scratch/validate.py
import os
import json
import re
from bs4 import BeautifulSoup

workspace_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
posts_json_path = os.path.join(workspace_dir, "data", "posts.json")
pages_json_path = os.path.join(workspace_dir, "data", "pages.json")

def validate_pages_json():
    print("--- Validating pages.json ---")
    if not os.path.exists(pages_json_path):
        print("FAIL: pages.json does not exist!")
        return None
    try:
        with open(pages_json_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
        print(f"SUCCESS: pages.json parsed correctly. Total pages: {len(data)}")
        return data
    except Exception as e:
        print(f"FAIL: Error parsing pages.json: {e}")
        return None

def validate_posts_json():
    print("--- Validating posts.json ---")
    if not os.path.exists(posts_json_path):
        print("FAIL: posts.json does not exist!")
        return None
    try:
        with open(posts_json_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
        print(f"SUCCESS: posts.json parsed correctly. Total posts: {len(data)}")
        return data
    except Exception as e:
        print(f"FAIL: Error parsing posts.json: {e}")
        return None

def check_file_exists(base_dir, relative_path, current_file_path, posts_data=None):
    # Resolve the path relative to the directory containing current_file_path
    file_dir = os.path.dirname(current_file_path)
    # Handle absolute looking or site-root looking paths if any, but since they should be relative:
    # Remove query params or hashes
    clean_path = relative_path.split('?')[0].split('#')[0]
    if not clean_path:
        return True # Just a hash link or empty
    
    # If it is absolute (starts with http, https, mailto, etc.), ignore
    if clean_path.startswith(('http://', 'https://', 'mailto:', '//', 'tel:')):
        return True
        
    # Check for SPA dynamic routes resolved relative to base_dir
    resolved_full_path = os.path.normpath(os.path.join(file_dir, clean_path))
    rel_from_root = os.path.relpath(resolved_full_path, base_dir).replace('\\', '/')

    if rel_from_root.startswith('posts/'):
        parts = rel_from_root.split('/')
        if len(parts) >= 2:
            slug = parts[1]
            if posts_data and any(p.get("slug") == slug for p in posts_data):
                return True
    elif rel_from_root.startswith('pages/'):
        parts = rel_from_root.split('/')
        if len(parts) >= 2:
            slug = parts[1]
            pages_json = os.path.join(base_dir, "data", "pages.json")
            if os.path.exists(pages_json):
                with open(pages_json, 'r', encoding='utf-8') as f:
                    pages_data = json.load(f)
                    if any(p.get("slug") == slug for p in pages_data):
                        return True
    elif rel_from_root.startswith('tags/') or rel_from_root.startswith('authors/'):
        return True

    resolved_path = os.path.normpath(os.path.join(file_dir, clean_path))
    
    # Check if file exists, or if it's a directory, check if index.html exists inside it
    if os.path.exists(resolved_path):
        return True
    
    # Check if appending index.html helps (e.g. if link is href="../../posts/some-post/")
    if os.path.isdir(resolved_path) or resolved_path.endswith('/') or not os.path.splitext(resolved_path)[1]:
        alt_path = os.path.normpath(os.path.join(resolved_path, "index.html"))
        if os.path.exists(alt_path):
            return True
            
    return False

def validate_html_files(posts_data, pages_data):
    print("--- Validating Post and Page HTML files ---")
    all_ok = True
    
    # 1. Validate posts
    for post in posts_data:
        slug = post["slug"]
        html_file = os.path.join(workspace_dir, "data", "posts", f"{slug}.html")
        fake_file_path = os.path.join(workspace_dir, "posts", slug, "index.html")
        
        if not os.path.exists(html_file):
            print(f"FAIL: Post HTML file missing for slug: {slug} at {html_file}")
            all_ok = False
            continue
            
        with open(html_file, 'r', encoding='utf-8') as f:
            soup = BeautifulSoup(f.read(), 'html.parser')
            
        elements_to_check = []
        for tag in soup.find_all(['a', 'link', 'img', 'script', 'source']):
            url = None
            attr = None
            if tag.name in ('a', 'link'):
                url = tag.get('href')
                attr = 'href'
            elif tag.name in ('img', 'script', 'source'):
                url = tag.get('src') or tag.get('srcset')
                attr = 'src/srcset'
                
            if url:
                urls = []
                if tag.name == 'source' or (tag.name == 'img' and tag.has_attr('srcset')):
                    for part in url.split(','):
                        part = part.strip()
                        if part:
                            urls.append(part.split()[0])
                else:
                    urls = [url]
                    
                for u in urls:
                    elements_to_check.append((tag.name, attr, u))
                    
        broken_links = []
        for name, attr, val in elements_to_check:
            if not check_file_exists(workspace_dir, val, fake_file_path, posts_data):
                broken_links.append((name, attr, val))
                
        if broken_links:
            print(f"FAIL: data/posts/{slug}.html has broken links:")
            for name, attr, val in broken_links:
                print(f"  - <{name} {attr}=\"{val}\"> is broken!")
            all_ok = False
            
    # 2. Validate pages
    for page in pages_data:
        slug = page["slug"]
        html_file = os.path.join(workspace_dir, "data", "pages", f"{slug}.html")
        fake_file_path = os.path.join(workspace_dir, "pages", slug, "index.html")
        
        if not os.path.exists(html_file):
            print(f"FAIL: Page HTML file missing for slug: {slug} at {html_file}")
            all_ok = False
            continue
            
        with open(html_file, 'r', encoding='utf-8') as f:
            soup = BeautifulSoup(f.read(), 'html.parser')
            
        elements_to_check = []
        for tag in soup.find_all(['a', 'link', 'img', 'script', 'source']):
            url = None
            attr = None
            if tag.name in ('a', 'link'):
                url = tag.get('href')
                attr = 'href'
            elif tag.name in ('img', 'script', 'source'):
                url = tag.get('src') or tag.get('srcset')
                attr = 'src/srcset'
                
            if url:
                urls = []
                if tag.name == 'source' or (tag.name == 'img' and tag.has_attr('srcset')):
                    for part in url.split(','):
                        part = part.strip()
                        if part:
                            urls.append(part.split()[0])
                else:
                    urls = [url]
                    
                for u in urls:
                    elements_to_check.append((tag.name, attr, u))
                    
        broken_links = []
        for name, attr, val in elements_to_check:
            if not check_file_exists(workspace_dir, val, fake_file_path, posts_data):
                broken_links.append((name, attr, val))
                
        if broken_links:
            print(f"FAIL: data/pages/{slug}.html has broken links:")
            for name, attr, val in broken_links:
                print(f"  - <{name} {attr}=\"{val}\"> is broken!")
            all_ok = False
            
    if all_ok:
        print("SUCCESS: All post and page HTML files have valid relative resource and link paths.")
    return all_ok

def validate_main_pages(posts_data=None):
    print("--- Validating Main HTML Pages ---")
    all_ok = True
    main_files = [
        os.path.join(workspace_dir, "index.html"),
        os.path.join(workspace_dir, "404.html")
    ]
    
    # Add paginated pages and tag pages
    for root, dirs, files in os.walk(workspace_dir):
        # skip posts and some system/data directories
        if any(p in root for p in ['/posts/', '/.git/', '/scratch/', 'data/posts', 'data/pages']):
            continue
        for f in files:
            if f.endswith('.html'):
                main_files.append(os.path.join(root, f))
                
    main_files = list(set(main_files)) # deduplicate
    print(f"Checking {len(main_files)} html files...")
    
    for html_file in main_files:
        if not os.path.exists(html_file):
            continue
        with open(html_file, 'r', encoding='utf-8') as f:
            soup = BeautifulSoup(f.read(), 'html.parser')
            
        elements_to_check = []
        for tag in soup.find_all(['a', 'link', 'img', 'script']):
            url = tag.get('href') or tag.get('src')
            attr = 'href' if tag.get('href') else 'src'
            if url:
                elements_to_check.append((tag.name, attr, url))
                
        broken_links = []
        for name, attr, val in elements_to_check:
            if not check_file_exists(workspace_dir, val, html_file, posts_data):
                broken_links.append((name, attr, val))
                
        # Report
        rel_path = os.path.relpath(html_file, workspace_dir)
        if broken_links:
            print(f"FAIL: {rel_path} has broken links:")
            for name, attr, val in broken_links:
                print(f"  - <{name} {attr}=\"{val}\"> is broken!")
            all_ok = False
            
    if all_ok:
        print("SUCCESS: All main pages have valid links/resources.")
    return all_ok

if __name__ == "__main__":
    posts = validate_posts_json()
    pages = validate_pages_json()
    if posts and pages:
        html_ok = validate_html_files(posts, pages)
        main_ok = validate_main_pages(posts)
        if html_ok and main_ok:
            print("\n*** ALL CHECKS PASSED SUCCESSFULLY! ***")
        else:
            print("\n*** SOME CHECKS FAILED! ***")
    else:
        print("Validation stopped due to metadata file loading failure.")

File: scripts/validate.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const workspaceDir = path.resolve(__dirname, '..');

const postsJsonPath = path.join(workspaceDir, 'data', 'posts.json');
const pagesJsonPath = path.join(workspaceDir, 'data', 'pages.json');

function validateJsonFile(filePath, name) {
  console.log(`--- Validating ${name} ---`);
  if (!fs.existsSync(filePath)) {
    console.error(`FAIL: ${name} does not exist at ${filePath}`);
    return null;
  }
  try {
    const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    console.log(`SUCCESS: ${name} parsed correctly. Total items: ${data.length}`);
    return data;
  } catch (err) {
    console.error(`FAIL: Error parsing ${name}:`, err.message);
    return null;
  }
}

function checkFileExists(relPath, currentFilePath, postsData) {
  // Strip query parameters or hashes
  let cleanPath = relPath.split('?')[0].split('#')[0];
  if (!cleanPath) return true; // anchor link

  // External links, mailto, etc.
  if (cleanPath.startsWith('http://') || cleanPath.startsWith('https://') || cleanPath.startsWith('mailto:') || cleanPath.startsWith('//') || cleanPath.startsWith('tel:')) {
    return true;
  }

  // Resolve path relative to directory of current file
  const currentDir = path.dirname(currentFilePath);

  // If path starts with site root '/', resolve relative to workspaceDir
  let resolvedPath;
  if (cleanPath.startsWith('/')) {
    resolvedPath = path.normalize(path.join(workspaceDir, cleanPath));
  } else {
    resolvedPath = path.normalize(path.join(currentDir, cleanPath));
  }

  const relFromRoot = path.relative(workspaceDir, resolvedPath).replace(/\\/g, '/');

  // Check SPA routes
  if (relFromRoot.startsWith('posts/')) {
    const slug = relFromRoot.split('/')[1];
    if (postsData && postsData.some(p => p.slug === slug)) {
      return true;
    }
  } else if (relFromRoot.startsWith('pages/')) {
    const slug = relFromRoot.split('/')[1];
    if (fs.existsSync(pagesJsonPath)) {
      try {
        const pagesData = JSON.parse(fs.readFileSync(pagesJsonPath, 'utf8'));
        if (pagesData.some(p => p.slug === slug)) {
          return true;
        }
      } catch (e) {}
    }
  } else if (relFromRoot.startsWith('tags/') || relFromRoot.startsWith('authors/')) {
    return true;
  }

  if (fs.existsSync(resolvedPath)) {
    return true;
  }

  // Directory fallback: if directory, check index.html inside it
  if (fs.existsSync(resolvedPath) && fs.statSync(resolvedPath).isDirectory()) {
    const indexPath = path.join(resolvedPath, 'index.html');
    if (fs.existsSync(indexPath)) {
      return true;
    }
  }

  // Fallback for relative folders
  if (!path.extname(resolvedPath)) {
    const altPath = path.join(resolvedPath, 'index.html');
    if (fs.existsSync(altPath)) {
      return true;
    }
  }

  return false;
}

// Simple parser to extract src, href, and srcset attributes
function extractUrls(htmlContent) {
  const urls = [];
  
  // 1. Match href="..."
  const hrefRegex = /href=["']([^"']+)["']/g;
  let match;
  while ((match = hrefRegex.exec(htmlContent)) !== null) {
    urls.push({ tag: 'link/a', attr: 'href', val: match[1] });
  }

  // 2. Match src="..."
  const srcRegex = /src=["']([^"']+)["']/g;
  while ((match = srcRegex.exec(htmlContent)) !== null) {
    urls.push({ tag: 'img/script', attr: 'src', val: match[1] });
  }

  // 3. Match srcset="..."
  const srcsetRegex = /srcset=["']([^"']+)["']/g;
  while ((match = srcsetRegex.exec(htmlContent)) !== null) {
    const srcsetValue = match[1];
    // srcset has comma-separated list of image-url width-descriptor
    srcsetValue.split(',').forEach(part => {
      const trimmed = part.trim();
      if (trimmed) {
        const urlPart = trimmed.split(/\s+/)[0];
        if (urlPart) {
          urls.push({ tag: 'source/img', attr: 'srcset', val: urlPart });
        }
      }
    });
  }

  return urls;
}

function validateHtmlFiles(postsData, pagesData) {
  console.log('--- Validating Compiled HTML Files ---');
  let allOk = true;

  // 1. Validate posts
  for (const post of postsData) {
    const slug = post.slug;
    const htmlFile = path.join(workspaceDir, 'content', 'posts', `${slug}.html`);
    // Simulated path context as seen in the SPA
    const fakeFilePath = path.join(workspaceDir, 'posts', slug, 'index.html');

    if (!fs.existsSync(htmlFile)) {
      console.error(`FAIL: Post HTML missing for slug: ${slug} at ${htmlFile}`);
      allOk = false;
      continue;
    }

    const htmlContent = fs.readFileSync(htmlFile, 'utf8');
    const urls = extractUrls(htmlContent);
    const broken = [];

    for (const item of urls) {
      if (!checkFileExists(item.val, fakeFilePath, postsData)) {
        broken.push(item);
      }
    }

    if (broken.length > 0) {
      console.error(`FAIL: content/posts/${slug}.html has broken links/resources:`);
      broken.forEach(item => {
        console.error(`  - [${item.tag}] ${item.attr}="${item.val}"`);
      });
      allOk = false;
    }
  }

  // 2. Validate pages
  for (const page of pagesData) {
    const slug = page.slug;
    const htmlFile = path.join(workspaceDir, 'content', 'pages', `${slug}.html`);
    const fakeFilePath = path.join(workspaceDir, 'pages', slug, 'index.html');

    if (!fs.existsSync(htmlFile)) {
      console.error(`FAIL: Page HTML missing for slug: ${slug} at ${htmlFile}`);
      allOk = false;
      continue;
    }

    const htmlContent = fs.readFileSync(htmlFile, 'utf8');
    const urls = extractUrls(htmlContent);
    const broken = [];

    for (const item of urls) {
      if (!checkFileExists(item.val, fakeFilePath, postsData)) {
        broken.push(item);
      }
    }

    if (broken.length > 0) {
      console.error(`FAIL: content/pages/${slug}.html has broken links/resources:`);
      broken.forEach(item => {
        console.error(`  - [${item.tag}] ${item.attr}="${item.val}"`);
      });
      allOk = false;
    }
  }

  if (allOk) {
    console.log('SUCCESS: All posts and pages have valid relative resource and link paths.');
  }
  return allOk;
}

function validateMainPages(postsData) {
  console.log('--- Validating Main HTML Pages ---');
  let allOk = true;

  const mainFiles = [
    path.join(workspaceDir, 'index.html'),
    path.join(workspaceDir, '404.html')
  ];

  for (const htmlFile of mainFiles) {
    if (!fs.existsSync(htmlFile)) {
      console.error(`FAIL: Main file missing: ${htmlFile}`);
      allOk = false;
      continue;
    }

    const htmlContent = fs.readFileSync(htmlFile, 'utf8');
    const urls = extractUrls(htmlContent);
    const broken = [];

    for (const item of urls) {
      if (!checkFileExists(item.val, htmlFile, postsData)) {
        broken.push(item);
      }
    }

    const relPath = path.relative(workspaceDir, htmlFile);
    if (broken.length > 0) {
      console.error(`FAIL: ${relPath} has broken links/resources:`);
      broken.forEach(item => {
        console.error(`  - [${item.tag}] ${item.attr}="${item.val}"`);
      });
      allOk = false;
    }
  }

  if (allOk) {
    console.log('SUCCESS: All main pages have valid links/resources.');
  }
  return allOk;
}

const posts = validateJsonFile(postsJsonPath, 'posts.json');
const pages = validateJsonFile(pagesJsonPath, 'pages.json');

if (posts && pages) {
  const htmlOk = validateHtmlFiles(posts, pages);
  const mainOk = validateMainPages(posts);

  if (htmlOk && mainOk) {
    console.log('\n*** ALL SANITY CHECKS PASSED SUCCESSFULLY! ***');
    process.exit(0);
  } else {
    console.error('\n*** SOME SANITY CHECKS FAILED! ***');
    process.exit(1);
  }
} else {
  console.error('Validation stopped due to metadata file loading failure.');
  process.exit(1);
}
